super_editor/lib/src/default_editor/multi_node_editing.dart

```diff
@@ -2,39 +2,423 @@ import 'dart:math';
+        )
+      ]);
+
+      newSelection = DocumentSelection.collapsed(
+        position: DocumentPosition(
+          nodeId: selectedNodeId,
+          nodePosition: newNode.endPosition,
+        ),
+      );
+    } else {
+      // Split the paragraph and inset image in between.
+      final textBefore = selectedNode.text.copyText(0, paragraphPosition.offset);
+      final textAfter = selectedNode.text.copyText(paragraphPosition.offset);
+
+      final newParagraph = ParagraphNode(id: Editor.createNodeId(), text: textAfter);
+
+      selectedNode.text = textBefore;
+      document
+        ..insertNodeAfter(existingNode: selectedNode, newNode: newNode)
+        ..insertNodeAfter(existingNode: newNode, newNode: newParagraph);
+      executor.logChanges([
+        DocumentEdit(
+          NodeChangeEvent(selectedNodeId),
+        ),
+        DocumentEdit(
+          NodeInsertedEvent(newNode.id, document.getNodeIndexById(newNode.id)),
+        ),
+        DocumentEdit(
+          NodeInsertedEvent(newParagraph.id, document.getNodeIndexById(newParagraph.id)),
+        ),
+      ]);
+
+      newSelection = DocumentSelection.collapsed(
+        position: DocumentPosition(
+          nodeId: newParagraph.id,
+          nodePosition: newParagraph.beginningPosition,
+        ),
+      );
+    }
+
+    executor.executeCommand(ChangeSelectionCommand(
+      newSelection,
+      SelectionChangeType.insertContent,
+      SelectionReason.userInteraction,
+    ));
+  }
+}
+
+class MoveNodeRequest implements EditRequest {
+  const MoveNodeRequest({
+    required this.nodeId,
+    required this.newIndex,
+  });
+
+  final String nodeId;
+  final int newIndex;
+}
+
+class MoveNodeCommand extends EditCommand {
+  MoveNodeCommand({
+    required this.nodeId,
+    required this.newIndex,
+  });
+
+  final String nodeId;
+  final int newIndex;
+
+  @override
+  void execute(EditContext context, CommandExecutor executor) {
+    final document = context.find<MutableDocument>(Editor.documentKey);
+
+    // Log all the move changes that will happen when we move the target node
+    // elsewhere in the document.
+    final nodeMoveEvents = <DocumentEdit>[];
+    final targetNodeIndex = document.getNodeIndexById(nodeId);
+    final startIndex = min(targetNodeIndex, newIndex);
+    final endIndex = max(targetNodeIndex, newIndex);
+    for (int i = startIndex; i <= endIndex; i += 1) {
+      if (i == targetNodeIndex) {
+        // This is the node that we care about moving.
+        nodeMoveEvents.add(
+          DocumentEdit(
+            NodeMovedEvent(nodeId: nodeId, from: targetNodeIndex, to: newIndex),
+          ),
+        );
+        continue;
+      }
+
+      // This is a node that got moved up by one as a consequence of moving
+      // the target node.
+      nodeMoveEvents.add(
+        DocumentEdit(
+          NodeMovedEvent(nodeId: document.getNodeAt(i)!.id, from: i, to: i - 1),
```

Comment:
While debugging and issue on our side with moving nodes, I noticed that we are decrementing instead of incrementing the `to` index. When moving a node to index 0, it would define `to` as -1.
```suggestion 
          NodeMovedEvent(nodeId: document.getNodeAt(i)!.id, from: i, to: i + 1),
```

Reply:
I think the original behavior is correct, and I don't think it should be possible for this situation to impact the 0'th node. Did you run into that in real execution?

With the 0'th node being the very first node in the list, the only way the 0'th node would be impacted by this loop is if that's the node that's being moved somewhere else. If it's being moved somewhere else, then it followed the if-statement above, rather than the adjustment you commented on.

When a node moves to a new location in the list, it causes all other nodes between those two positions to move up by one, hence the `- 1`.

Reply:
I wrote a small test where we can check this:
```dart
import 'package:flutter_test/flutter_test.dart';
import 'package:super_editor/super_editor.dart';

void main() {
  test('move a node', () {
    final document = MutableDocument(
      nodes: [
        ParagraphNode(
          id: 'p1',
          text: AttributedText(text: 'Paragraph 1'),
        ),
        ParagraphNode(
          id: 'p2',
          text: AttributedText(text: 'Paragraph 2'),
        ),
        ParagraphNode(
          id: 'p3',
          text: AttributedText(text: 'Paragraph 3'),
        ),
      ],
    );
    final editor = Editor(
      editables: {Editor.documentKey: document},
      requestHandlers: [
        (request) => request is MoveNodeRequest
            ? MoveNodeCommand(
                nodeId: request.nodeId,
                newIndex: request.newIndex,
              )
            : null,
      ],
    );
    const request = MoveNodeRequest(
      nodeId: 'p3',
      newIndex: 0,
    );
    final changeLog = <DocumentChangeLog>[];
    document.addListener(changeLog.add);

    editor.execute([request]);

    expect(
      document.nodes.map((node) => node.id),
      ['p3', 'p1', 'p2'],
    );

    print(changeLog.map((e) => e.changes));
    expect(changeLog.length, 1);
    expect(
      changeLog.first.changes,
      [
        const NodeMovedEvent(nodeId: 'p1', from: 0, to: 1),
        const NodeMovedEvent(nodeId: 'p2', from: 1, to: 2),
        const NodeMovedEvent(nodeId: 'p3', from: 2, to: 0),
      ],
    );
  });
}

```

The last `expect` where we check the change log fails. The changes reported with the current code are:
```dart
    expect(
      changeLog.first.changes,
      [
        const NodeMovedEvent(nodeId: 'p1', from: 0, to: -1),
        const NodeMovedEvent(nodeId: 'p2', from: 1, to: 0),
        const NodeMovedEvent(nodeId: 'p3', from: 2, to: 0),
      ],
    );
```